gem/optimise.py

```diff
@@ -413,3 +474,403 @@ def aggressive_unroll(expression):
+                if key == 'const' or key == 'other':
+                    continue
+                for factor in factors:
+                    if factor not in gem_int:
+                        gem_int[factor] = counter
+                        int_gem[counter] = factor
+                        counter += 1
+        if counter == 0:
+            return tuple()
+        # add connections (list of tuples)
+        connections = []
+        for summand in self.rep:
+            connection = []
+            for key, factors in iteritems(summand):
+                if key == 'const' or key == 'other':
+                    continue
+                connection.extend([gem_int[factor] for factor in factors])
+            connections.append(tuple(connection))
+
+        # set up the ILP
+        import pulp as ilp
+        ilp_prob = ilp.LpProblem('gem factorise', ilp.LpMinimize)
+        ilp_var = ilp.LpVariable.dicts('node', int_gem.keys(), 0, 1, ilp.LpBinary)
+
+        # Objective function
+        # Minimise number of factors to pull. If same number, favour factor with larger extent
+        big = 10000000  # some arbitrary big number
+        ilp_prob += ilp.lpSum(ilp_var[i] * (big - self.factor_extent(int_gem[i])) for i in int_gem)
+
+        # constraints
+        for connection in connections:
+            ilp_prob += ilp.lpSum(ilp_var[i] for i in connection) >= 1
+
+        ilp_prob.solve()
+        if ilp_prob.status != 1:
+            raise AssertionError("Something bad happened during ILP")
+
+        optimal_factors = [factor for factor, number in iteritems(gem_int) if ilp_var[number].value() == 1]
+        other_factors = [factor for factor, number in iteritems(gem_int) if ilp_var[number].value() == 0]
+        # TODO: investigate effects of sorting these two lists of factors
+        optimal_factors = sorted(optimal_factors, key=lambda f: self.factor_extent(f), reverse=True)
+        other_factors = sorted(other_factors, key=lambda f: self.factor_extent(f), reverse=True)
+        # Sequence dictating order of factorisation
+        factors_seq = optimal_factors + other_factors
+        return factors_seq
+
+    def factorise_key(self, key):
+        """
+        Factorise common factors that have a particular :param key
+        """
+        # TODO: Add a forbidden list of factors
+        if len(self.rep) < 2:
+            return
+        counter = OrderedDict()
+        for summand in self.rep:
+            if key not in summand:
+                continue
+            for factor in summand[key]:
+                counter.setdefault(factor, 0)
+                counter[factor] += 1
+        if max(counter.values()) < 2:
+            return
+        saved_flops = OrderedDict((((count_flop(factor) + 1)) * (count - 1), factor)
+                                  for (factor, count) in iteritems(counter))
+        mcf = saved_flops[max(saved_flops.keys())]  # most common factor
```

Comment:
This is really strange way to write a loop that just does `max`.  Why do you need this giant temporary?